File: apps/motoboy/app/src/test/kotlin/com/drogsantoantonio/farmadelivery/presentation/ui/screens/RouteStopDisplayTest.kt
package com.drogsantoantonio.farmadelivery.presentation.ui.screens

import com.drogsantoantonio.farmadelivery.data.models.RouteStopDeliveryDto
import com.drogsantoantonio.farmadelivery.data.models.RouteStopDto
import kotlin.test.Test
import kotlin.test.assertEquals

class RouteStopDisplayTest {
  @Test
  fun `uses delivery public code as route stop title when available`() {
    assertEquals("FD-123", routeStopTitle(stop(delivery = delivery(publicCode = "FD-123"))))
  }

  @Test
  fun `uses store daily number as route stop title when available`() {
    assertEquals("N. dia 009", routeStopTitle(stop(delivery = delivery(publicCode = "AST-20260517-009", storeDailyNumber = 9))))
  }

  @Test
  fun `falls back to route stop type when delivery context is unavailable`() {
    assertEquals("Entrega", routeStopTitle(stop(delivery = null)))
  }

  @Test
  fun `shows customer and delivery status as route stop details`() {
    assertEquals("Maria - Em rota", routeStopDetails(stop(delivery = delivery(customer = "Maria", status = "EM_ROTA"))))
  }

  @Test
  fun `keeps full public code in route stop details when title uses daily number`() {
    assertEquals(
      "AST-20260517-009 - Maria - Em rota",
      routeStopDetails(stop(delivery = delivery(publicCode = "AST-20260517-009", storeDailyNumber = 9, customer = "Maria", status = "EM_ROTA"))),
    )
  }

  @Test
  fun `falls back to delivery status when customer is blank`() {
    assertEquals("Disponivel", routeStopDetails(stop(delivery = delivery(customer = "  ", status = "AGUARDANDO_MOTOBOY"))))
  }

  @Test
  fun `omits route stop details when delivery context is unavailable`() {
    assertEquals(null, routeStopDetails(stop(delivery = null)))
  }

  private fun stop(delivery: RouteStopDeliveryDto?): RouteStopDto {
    return RouteStopDto(
      id = "stop-1",
      sequence = 1,
      type = "ENTREGA",
      status = "PENDENTE",
      address = "Rua Teste",
      delivery = delivery,
    )
  }

  private fun delivery(
    publicCode: String = "FD-001",
    storeDailyNumber: Int? = null,
    customer: String = "Cliente Teste",
    status: String = "AGUARDANDO_MOTOBOY",
  ): RouteStopDeliveryDto {
    return RouteStopDeliveryDto(
      id = "delivery-1",
      publicCode = publicCode,
      storeDailyNumber = storeDailyNumber,
      customer = customer,
      status = status,
    )
  }
}

File: apps/motoboy-pwa/src/routeStopDisplay.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import { routeStopDetails, routeStopTitle } from "./routeStopDisplay";
import type { CourierRoute } from "./types";

test("uses delivery public code as route stop title when available", () => {
  assert.equal(routeStopTitle(stop({ publicCode: "FD-123" })), "FD-123");
});

test("uses store daily number as route stop title when available", () => {
  assert.equal(routeStopTitle(stop({ publicCode: "AST-20260517-009", storeDailyNumber: 9 })), "N. dia 009");
});

test("falls back to route stop type when delivery context is unavailable", () => {
  assert.equal(routeStopTitle(stop(null)), "Entrega");
});

test("shows customer and delivery status as route stop details", () => {
  assert.equal(routeStopDetails(stop({ customer: "Maria", status: "EM_ROTA" })), "Maria - Em rota");
});

test("keeps full public code in route stop details when title uses daily number", () => {
  assert.equal(
    routeStopDetails(stop({ publicCode: "AST-20260517-009", storeDailyNumber: 9, customer: "Maria", status: "EM_ROTA" })),
    "AST-20260517-009 - Maria - Em rota",
  );
});

test("falls back to delivery status when customer is blank", () => {
  assert.equal(routeStopDetails(stop({ customer: "  ", status: "AGUARDANDO_MOTOBOY" })), "Disponivel");
});

test("omits route stop details when delivery context is unavailable", () => {
  assert.equal(routeStopDetails(stop(null)), null);
});

function stop(
  delivery: Partial<NonNullable<CourierRoute["stops"][number]["delivery"]>> | null,
): CourierRoute["stops"][number] {
  return {
    id: "stop-1",
    sequence: 1,
    type: "ENTREGA",
    status: "PENDENTE",
    address: "Rua Teste",
    latitude: null,
    longitude: null,
    earliestAt: null,
    completedAt: null,
    delivery: delivery
      ? {
          id: "delivery-1",
          publicCode: delivery.publicCode ?? "FD-001",
          storeDailyNumber: delivery.storeDailyNumber ?? null,
          customer: delivery.customer ?? "Cliente Teste",
          status: delivery.status ?? "AGUARDANDO_MOTOBOY",
        }
      : null,
  };
}
